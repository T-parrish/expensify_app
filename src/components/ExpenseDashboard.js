import React from 'react';

const ExpenseDashboard = () => (
	<div>
		<p>This is from my Dashboard component</p>
	</div>
);

export default ExpenseDashboard;