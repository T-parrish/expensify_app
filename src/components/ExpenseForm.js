import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),  
            calFocused: false,
            errorState: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calFocused: focused }))
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ errorState: 'Please specify description and amount'}))
        } else {
            this.setState(() => ({ errorState: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.errorState.length > 0 && <p className="form__error">{this.state.errorState}</p>}
                <input
                    type="text"
                    className="text-input"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <div className="form__sub">
                    <input
                        type="text"
                        className="text-input__amount"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} 
                        onDateChange={this.onDateChange} 
                        focused={this.state.calFocused} 
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </div>
                <textarea
                    placeholder="add a note for your expense"
                    className="text-area"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className="button--submit">Save Expense</button>
                </div>
            </form>
        )
    }
}