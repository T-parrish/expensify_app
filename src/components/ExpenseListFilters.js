import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate  } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused }))
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    };

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-sub">
                        <div className="input-group__item">
                            <input 
                                className="text-input text-input__filter"
                                type="text" 
                                placeholder="search expenses"
                                value={this.props.filters.text} 
                                onChange={this.onTextChange}
                            />
                        </div>
                        <div className="input-group__item">
                            <select
                                className="select"
                                value = {this.props.filters.sortBy}
                                onChange={this.onSortChange}
                            >
                                <option value = "date"> Date </option>
                                <option value = "amount"> Amount </option>
                            </select>
                        </div>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate = {this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            startDateId='start'
                            endDateId='end'
                            showClearDates={true}
                            numberOfMonths = {1}
                            isOutsideRange= {() => false}
                        />
                    </div>
                </div>
            </div> 
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

