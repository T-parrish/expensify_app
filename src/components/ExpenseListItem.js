import React from 'react'
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <h3>{description} - <Link to={`/edit/${id}`}> edit </Link> </h3>
        <p>
            {numeral(amount / 100).format('$0,0.00')} 
            - 
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);


export default ExpenseListItem