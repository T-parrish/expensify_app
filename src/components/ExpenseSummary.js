import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpenses = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <p> Viewing {expenseCount} {expenseWord} totalling {formattedExpenses}  </p>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary)