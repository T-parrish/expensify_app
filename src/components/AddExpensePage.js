import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startAddExpense(expense);
		this.props.history.push('/');
	};
	render() {
		return (
			<div>
				<p>Add Expense</p>
				<ExpenseForm onSubmit={this.onSubmit}/>
			</div>
		);
	};
};

const mapDispatchToProps = (dispatch) => ({
	startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);