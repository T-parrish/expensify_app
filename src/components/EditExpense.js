import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	};

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm
						expense = {this.props.expense}
						onSubmit= {this.onSubmit}
					/>
					<button className="button_three" onClick={this.onRemove}> Remove </button>
				</div>
			</div>
		);
	};	
};

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find((expense) => {
			return expense.id === props.match.params.id;
		})
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);

