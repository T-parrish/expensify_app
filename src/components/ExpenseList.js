import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpense from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p> No Expenses </p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense}/>
                })
            )
        }
        
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filters)
    };
};


export default connect(mapStateToProps)(ExpenseList);

// {props.filters.text}
// {props.expenses.length}


