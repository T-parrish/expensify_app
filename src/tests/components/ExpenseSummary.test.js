import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should correctly render Expense totals with one expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={456} />)
    expect(wrapper).toMatchSnapshot();
}); 

test('should correctly render Expense totals with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={5367} />)
    expect(wrapper).toMatchSnapshot();
});
