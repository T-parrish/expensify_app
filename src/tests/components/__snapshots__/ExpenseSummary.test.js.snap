// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Expense totals with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
       Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $53.67
      </span>
         
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button_two"
        replace={false}
        to="/create"
      >
         Create Expense 
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render Expense totals with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
       Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $4.56
      </span>
         
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button_two"
        replace={false}
        to="/create"
      >
         Create Expense 
      </Link>
    </div>
  </div>
</div>
`;
